### PR TITLE
Add Dirichlet Evidential Uncertainty Quantification

### PR DESCRIPTION
**Pull request by cjmcgill**
## Description
Creates an option for `--uncertainty_method dirichlet`. This option allows for the calculation of dirichlet evidential uncertainty from predictions. Previously, classification methods trained with dirichlet loss function only provided the expected probability (excluding pure uncertainty). Now using this option, the dirichlet evidential uncertainty can be obtained in the prediction output.

Because this uncertainty output is not a confidence level, the calibration methods and evaluation methods available for calibration jobs normally are not available when `--uncertainty_method dirichlet`. However, calibration and evaluation can still be performed when the job is run with `--uncertainty_method classification` so that it's working with predicted confidence.

## Relevant issues
Solves issue in #413.
Supercedes PR #421

### CODE REVIEW

**Comment by cjmcgill** [comment]
@muammar I'd appreciate it if you could try out this branch and see if it matches your expectations.

**Comment by muammar** [comment]
The new modifications look good. I will try them this week. Thanks.

**Comment by muammar** [comment]
Everything worked correctly for my use case. Thanks for your help.

File: chemprop/uncertainty/uncertainty_calibrator.py [hunk 1]
```diff
@@ -86,6 +86,13 @@ def raise_argument_errors(self):
                     multiclass dataset types, their outputs are not confidences and are not \
                     compatible with any implemented calibration methods for classification."
             )
+        if self.uncertainty_method == "dirichlet":
+            raise NotImplementedError(
```
Comment (JacksonBurns) [inline chemprop/uncertainty/uncertainty_calibrator.py:90]: This should not be a `NotImplementedError`, but something more like a custom `InvalidConfigurationError` or I suggest just a plain old `RuntimeError`. See the [Python docs](https://docs.python.org/3/library/exceptions.html#NotImplementedError) - `NotImplementedError` is only for base classes with methods that are required to be overridden.

File: chemprop/uncertainty/uncertainty_evaluator.py [hunk 2]
```diff
@@ -51,6 +51,13 @@ def raise_argument_errors(self):
                     multiclass dataset types, their outputs are not confidences and are not \
                     compatible with any implemented evaluation methods for classification."
             )
+        if self.uncertainty_method == "dirichlet":
+            raise NotImplementedError(
```
Comment (JacksonBurns) [inline chemprop/uncertainty/uncertainty_evaluator.py:55]: Same comment as above.

File: chemprop/uncertainty/uncertainty_predictor.py [hunk 3]
```diff
@@ -1250,6 +1251,142 @@ def get_uncal_output(self):
...
+
+    def raise_argument_errors(self):
+        super().raise_argument_errors()
+        if self.loss_function != "dirichlet":
+            raise ValueError(
+                "In order to use Dirichlet evidential uncertainty, trained models must have used dirichlet loss function."
+            )
+        if self.dataset_type not in ["classification", "multiclass"]:
+            raise ValueError(
+                "Dirichlet evidential epistemic uncertainty is only compatible with classification dataset types."
```
Comment (JacksonBurns) [inline chemprop/uncertainty/uncertainty_predictor.py:1271]: Would be nice to see what we currently have with something like `f"Dirichlet ... types. Current dataset is of type {self.dataset_type}"`

**Review by JacksonBurns — CHANGES_REQUESTED** [review]
A couple small comments on the code itself. I have not validated the accuracy and am trusting the implementation and acceptance from @muammar.

**Comment by kevingreenman** [comment]
@cjmcgill thanks for updating the error message. Do you agree with @JacksonBurns on the `NotImplementedError`s?

**Comment by cjmcgill** [comment]
@kevingreenman @JacksonBurns I'm changing the `NotImplementedError` to `ValueError`. `ValueError` is what we've used for this kind of configuration/options error elsewhere and that would be more consistent.

I think that the custom error for configuration error is a good idea, but making that change would cover a lot of territory. There's many instances of ValueError for this and I think we're probably not the most interested in making that wide-ranging of a change in an unrelated PR.

**Review by kevingreenman — APPROVED** [review]

**Review by JacksonBurns — APPROVED** [review]
Looks good, thanks for addressing my comments. I agree with your thoughts on keeping this in scope as well.

**Pull request merged**